module/epochs/machine_account.go

```diff
@@ -0,0 +1,140 @@
+	sdkcrypto "github.com/onflow/flow-go-sdk/crypto"
+	"github.com/onflow/flow-go/model/bootstrap"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+var (
+	// Hard and soft balance limits for collection and consensus nodes.
+	// We will log a warning once for a soft limit, and will log an error
+	// in perpetuity for a hard limit.
+	// Taken from https://www.notion.so/dapperlabs/Machine-Account-f3c293593ea442a39614fcebf705a132
+
+	SoftMinBalanceLN cadence.UFix64
+	HardMinBalanceLN cadence.UFix64
+	SoftMinBalanceSN cadence.UFix64
+	HardMinBalanceSN cadence.UFix64
+)
+
+func init() {
+	var err error
+	SoftMinBalanceLN, err = cadence.NewUFix64("0.0025")
+	if err != nil {
+		panic(err)
+	}
+	HardMinBalanceLN, err = cadence.NewUFix64("0.001")
+	if err != nil {
+		panic(err)
+	}
+	SoftMinBalanceSN, err = cadence.NewUFix64("0.125")
+	if err != nil {
+		panic(err)
+	}
+	HardMinBalanceSN, err = cadence.NewUFix64("0.05")
+	if err != nil {
+		panic(err)
+	}
+}
+
+// CheckMachineAccountInfo checks a node machine account config, logging
+// anything noteworthy but not critical, and returning an error if the machine
+// account is not configured correctly, or the configuration cannot be checked.
+func CheckMachineAccountInfo(
+	log zerolog.Logger,
+	role flow.Role,
+	info bootstrap.NodeMachineAccountInfo,
+	account *sdk.Account,
+) error {
+
+	log.Debug().
+		Str("machine_account_address", info.Address).
+		Str("role", role.String()).
+		Msg("checking machine account configuration...")
+
+	if role != flow.RoleCollection && role != flow.RoleConsensus {
+		return fmt.Errorf("invalid role (%s) must be one of [collection, consensus]", role.String())
+	}
+
+	address := info.FlowAddress()
+	if address == flow.EmptyAddress {
+		return fmt.Errorf("could not parse machine account address: %s", info.Address)
+	}
+
+	privKey, err := sdkcrypto.DecodePrivateKey(info.SigningAlgorithm, info.EncodedPrivateKey)
+	if err != nil {
+		return fmt.Errorf("could not decode machine account private key: %w", err)
+	}
+
+	// FIRST - check the local account info independently
+	if info.HashAlgorithm != bootstrap.DefaultMachineAccountHashAlgo {
+		log.Warn().Msgf("non-standard hash algo (expected %s, got %s)", bootstrap.DefaultMachineAccountHashAlgo, info.HashAlgorithm.String())
+	}
+	if info.SigningAlgorithm != bootstrap.DefaultMachineAccountSignAlgo {
+		log.Warn().Msgf("non-standard signing algo (expected %s, got %s)", bootstrap.DefaultMachineAccountSignAlgo, info.SigningAlgorithm.String())
+	}
+	if info.KeyIndex != bootstrap.DefaultMachineAccountKeyIndex {
+		log.Warn().Msgf("non-standard key index (expected %d, got %d)", bootstrap.DefaultMachineAccountKeyIndex, info.KeyIndex)
+	}
+
+	// SECOND - compare the local account info to the on-chain account
+	if !bytes.Equal(account.Address.Bytes(), address.Bytes()) {
+		return fmt.Errorf("machine account address mismatch between local (%s) and on-chain (%s)", address, account.Address)
+	}
+	if len(account.Keys) < 1 {
```

Comment:
Do we require that only 1 key exists or atleast 1 ? If it's the latter should we loop through them and do the verification from lines 96-112 on all the keys?

Reply:
We require that exactly one key exists which corresponds to the `NodeMachineAccountInfo` file. The account is allowed to have more keys (this is supported by Flow Port and the transactions, though it's not part of the supported process), and we don't care what those extra keys are here. 

Our tooling always makes the machine account key the first key so this conditional is making that assumption, but since the `KeyIndex` is part of the `NodeMachineAccountInfo`, we should probably be more flexible here, like:
```go
if len(account.Keys) <= machineAccountInfo.KeyIndex { ... }
```